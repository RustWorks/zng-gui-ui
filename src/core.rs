pub mod animation;
pub mod app;
pub mod context;
pub mod event;
pub mod events;
pub mod focus;
pub mod font;
pub mod frame;
pub mod profiler;
pub mod types;
pub mod var;
pub mod window;

use crate::impl_ui_node;
use context::{LazyStateMap, WidgetContext};
use frame::FrameBuilder;
use types::{LayoutSize, WidgetId};

/// An Ui tree node.
pub trait UiNode: 'static {
    /// Called every time the node is plugged in an Ui tree.
    fn init(&mut self, ctx: &mut WidgetContext);

    /// Called every time the node is unplugged from an Ui tree.
    fn deinit(&mut self, ctx: &mut WidgetContext);

    /// Called every time a low pressure event update happens.
    ///
    /// # Event Pressure
    /// See [update_hp] for more information about event pressure rate.
    fn update(&mut self, ctx: &mut WidgetContext);

    /// Called every time a high pressure event update happens.
    ///
    /// # Event Pressure
    /// Some events occur alot more times then others, for performance reasons this
    /// event source may choose to be propagated in the this hight pressure lane.
    ///
    /// Event sources that are high pressure mention this in their documentation.
    fn update_hp(&mut self, ctx: &mut WidgetContext);

    /// Called every time a layout update is needed.
    ///
    /// # Arguments
    /// * `available_size`: The total available size for the node. Can contain positive infinity to
    /// indicate the parent will accommodate any size.
    ///
    /// # Return
    /// Must return the nodes desired size. Must not contain infinity or NaN.
    fn measure(&mut self, available_size: LayoutSize) -> LayoutSize;

    /// Called every time a layout update is needed, after [measure].
    ///
    /// # Arguments
    /// * `final_size`: The size the parent node reserved for the node. Must reposition its contents
    /// to fit this size. The value does not contain infinity or NaNs.
    fn arrange(&mut self, final_size: LayoutSize);

    /// Called every time a new frame must be rendered.
    ///
    /// # Arguments
    /// * `frame`: Contains the next frame draw instructions.
    fn render(&self, frame: &mut FrameBuilder);

    /// Box this component, unless it is already `Box<dyn UiNode>`.
    fn boxed(self) -> Box<dyn UiNode>
    where
        Self: Sized + 'static,
    {
        Box::new(self)
    }
}

#[impl_ui_node(delegate: self.as_ref(), delegate_mut: self.as_mut())]
impl UiNode for Box<dyn UiNode> {
    fn boxed(self) -> Box<dyn UiNode> {
        self
    }
}

struct Widget<T: UiNode> {
    id: WidgetId,
    state: LazyStateMap,
    child: T,
}

#[impl_ui_node(child)]
impl<T: UiNode> UiNode for Widget<T> {
    fn init(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.init(ctx));
    }

    fn deinit(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.deinit(ctx));
    }

    fn update(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update(ctx));
    }

    fn update_hp(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update_hp(ctx));
    }

    fn render(&self, frame: &mut FrameBuilder) {
        frame.push_widget(self.id, &self.child);
    }
}

/// Creates a widget bondary.
pub fn widget(id: WidgetId, child: impl UiNode) -> impl UiNode {
    Widget {
        id,
        state: LazyStateMap::default(),
        child,
    }
}
