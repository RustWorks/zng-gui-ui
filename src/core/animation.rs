use crate::core::context::Vars;
use crate::core::var::VarValue;
use std::fmt::Debug;
use std::{ops::Mul, time::Duration};

/// A [variable](crate::core::var::Var) animation.
pub trait VarAnimation<T: VarValue> {
    /// Current animating value.
    fn get_step(&self, vars: &Vars) -> &T;

    /// Update animating value.
    fn update_step(&self, vars: &Vars) -> Option<&T>;

    /// Borrow end value.
    fn get_end(&self) -> &T;

    /// Stops the animation and returns  the end value.
    fn end(self) -> T;
}

/// Easing function value. Mostly between 0.0 and 1.0 but can over/undershot.
#[derive(Debug, Clone, Copy)]
pub struct EasingStep(pub f32);

impl EasingStep {
    /// Gets the scale value.
    #[inline]
    pub fn get(self) -> f32 {
        self.0
    }

    /// Flipped step.
    #[inline]
    pub fn flip(self) -> EasingStep {
        EasingStep(1.0 - self.0)
    }
}

/// Easing function time input. Is a value in the `0.0..=1.0` range.
#[derive(Debug, Clone, Copy)]
pub struct EasingTime(f32);

impl EasingTime {
    /// New easing time from calculated value.
    pub fn from_raw(raw: f32) -> EasingTime {
        EasingTime(raw.max(0.0).min(1.0))
    }

    /// New easing time from total `duration` and `elapsed` time.
    #[inline]
    pub fn new(duration: Duration, elapsed: Duration) -> EasingTime {
        if elapsed < duration {
            EasingTime(duration.as_secs_f32() / elapsed.as_secs_f32())
        } else {
            EasingTime(1.0)
        }
    }

    /// Gets the scale value.
    #[inline]
    pub fn get(self) -> f32 {
        self.0
    }

    /// Inverted time.
    #[inline]
    pub fn reverse(self) -> EasingTime {
        EasingTime(1.0 - self.0)
    }
}

/// A [VarValue] that is the final value of a transition animation and
/// can be multiplied by a [EasingStep] to get an intermediary value of the
/// same type.
pub trait EasingVarValue: VarValue + Mul<EasingStep, Output = Self> {}

impl<T: VarValue + Mul<EasingStep, Output = Self>> EasingVarValue for T {}

impl Mul<EasingStep> for f32 {
    type Output = Self;

    fn mul(self, rhs: EasingStep) -> f32 {
        self * rhs.0
    }
}

impl Mul<EasingStep> for f64 {
    type Output = Self;

    fn mul(self, rhs: EasingStep) -> f64 {
        self * rhs.0 as f64
    }
}

macro_rules! mul_easing_step_for_small_int {
    ($($ty:ty),+) => {$(
        impl Mul<EasingStep> for $ty {
            type Output = Self;

            #[inline]
            fn mul(self, rhs: EasingStep) -> Self {
                (self as f32 * rhs.0) as Self
            }
        }
    )+};
}

mul_easing_step_for_small_int!(u8, i8, u16, i16, u32, i32);

macro_rules! mul_easing_step_for_int  {
    ($($ty:ty),+) => {$(
        impl Mul<EasingStep> for $ty {
            type Output = Self;

            #[inline]
            fn mul(self, rhs: EasingStep) -> Self {
                (self as f64 * rhs.0 as f64) as Self
            }
        }
    )+};
}

mul_easing_step_for_int!(usize, isize, u64, i64, u128, i128);

/// Common easing functions.
///
/// See also: [`EasingFn`](EasingFn).
pub mod easing {
    use super::{EasingStep, EasingTime};
    use std::f32::consts::*;

    /// Applies the `ease_fn`.
    #[inline]
    pub fn ease_in(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        ease_fn(time)
    }

    /// Applies the `ease_fn` in reverse and flipped.
    #[inline]
    pub fn ease_out(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        ease_fn(time.reverse()).flip()
    }

    /// Applies `ease_in` for the first half then `[ease_out]` scaled to fit a single duration (1.0).
    pub fn ease_in_out(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        let time = EasingTime(time.get() * 2.0);
        let step = if time.get() < 1.0 {
            ease_in(ease_fn, time)
        } else {
            ease_out(ease_fn, time)
        };
        EasingStep(step.get() * 0.5)
    }

    /// Simple linear transition, no easing, no acceleration.
    #[inline]
    pub fn linear(time: EasingTime) -> EasingStep {
        EasingStep(time.get())
    }

    #[inline]
    pub fn quad(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t)
    }

    #[inline]
    pub fn cubic(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t)
    }

    #[inline]
    pub fn quart(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t * t)
    }

    #[inline]
    pub fn quint(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t * t * t)
    }

    #[inline]
    pub fn sine(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(1.0 - (t * FRAC_PI_2 * (1.0 - t)).sin())
    }

    #[inline]
    pub fn expo(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep((10.0 * (t - 1.0)).powf(2.0))
    }

    #[inline]
    pub fn circ(time: EasingTime) -> EasingStep {
        EasingStep(1.0 - (1.0 - time.get()).sqrt())
    }

    #[inline]
    pub fn back(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * (2.70158 * t - 1.70158))
    }

    #[inline]
    pub fn elastic(time: EasingTime) -> EasingStep {
        let t = time.get();
        let t2 = t * t;
        EasingStep(t2 * t2 * (t * PI * 4.5).sin())
    }

    #[inline]
    pub fn bounce(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep((6.0 * (t - 1.0)).powf(2.0) * (t * PI * 3.5).sin().abs())
    }
}

/// Common easing functions as an enum.
#[derive(Debug, Clone, Copy)]
pub enum EasingFn {
    Linear,
    Sine,
    Quad,
    Cubic,
    Quart,
    Quint,
    Expo,
    Circ,
    Back,
    Elastic,
    Bounce,
}

impl EasingFn {
    #[inline]
    pub fn ease_in(self, time: EasingTime) -> EasingStep {
        match self {
            EasingFn::Linear => easing::linear(time),
            EasingFn::Sine => easing::sine(time),
            EasingFn::Quad => easing::quad(time),
            EasingFn::Cubic => easing::cubic(time),
            EasingFn::Quart => easing::quad(time),
            EasingFn::Quint => easing::quint(time),
            EasingFn::Expo => easing::expo(time),
            EasingFn::Circ => easing::circ(time),
            EasingFn::Back => easing::back(time),
            EasingFn::Elastic => easing::elastic(time),
            EasingFn::Bounce => easing::bounce(time),
        }
    }

    #[inline]
    pub fn ease_out(self, time: EasingTime) -> EasingStep {
        easing::ease_out(|t| self.ease_in(t), time)
    }

    #[inline]
    pub fn ease_in_out(self, time: EasingTime) -> EasingStep {
        easing::ease_in_out(|t| self.ease_in(t), time)
    }
}
