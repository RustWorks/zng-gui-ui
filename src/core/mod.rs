#[macro_use]
mod macros;

mod focus;
mod font;
mod hittest;
mod keyboard;
mod mouse;
mod next_frame;
mod next_update;
mod ui;
mod ui_values;
mod window;

pub use focus::*;
pub use font::*;
pub use glutin::event::{ElementState, ModifiersState, MouseButton, ScanCode, VirtualKeyCode};
pub use glutin::window::CursorIcon;
pub use hittest::*;
pub use keyboard::*;
pub use mouse::*;
pub use next_frame::*;
pub use next_update::*;
pub use ui::*;
pub use ui_values::*;
pub use webrender::api::{
    ColorF, FontInstanceKey, FontKey, GradientStop, LayoutPoint, LayoutRect, LayoutSideOffsets, LayoutSize,
};
pub use webrender::euclid::{point2, size2};
pub use window::*;
