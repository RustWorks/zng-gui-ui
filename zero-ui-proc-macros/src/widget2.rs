use std::mem;

use proc_macro2::{Ident, TokenStream, TokenTree};
use quote::{quote, ToTokens};
use syn::{parse::Parse, spanned::Spanned, *};

use crate::{
    util::{self, parse_outer_attrs, Attributes, ErrorRecoverable, Errors},
    widget_util::{self, parse_remove, WgtProperty, WgtWhen},
};

pub fn expand(args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    // the widget mod declaration.
    let mod_ = parse_macro_input!(input as ItemMod);

    let uuid = util::uuid(&args); // full path to widget must be unique.

    if mod_.content.is_none() {
        let mut r = syn::Error::new(mod_.semi.span(), "only modules with inline content are supported")
            .to_compile_error()
            .to_token_stream();

        mod_.to_tokens(&mut r);

        return r.into();
    }

    let (_, items) = mod_.content.unwrap();

    // accumulate the most errors as possible before returning.
    let mut errors = Errors::default();

    let crate_core = util::crate_core();

    let vis = mod_.vis;
    let ident = mod_.ident;
    let attrs = mod_.attrs;
    let wgt_cfg = Attributes::new(attrs.clone()).cfg;

    // a `$crate` path to the widget module.
    let mod_path = match syn::parse::<ArgPath>(args) {
        Ok(a) => a.path,
        Err(e) => {
            errors.push_syn(e);
            quote! { $crate::missing_widget_path}
        }
    };

    let WidgetItems {
        uses,
        inherits,
        properties,
        intrinsic_fn,
        build_fn,
        others,
    } = WidgetItems::new(items, &mut errors);

    let mut intrinsic = quote!();

    for Inherit { attrs, path } in &inherits {
        intrinsic.extend(quote_spanned! {path.span()=>
            #(#attrs)*
            #path::__intrinsic__(__wgt__);
        });
    }

    if let Some(int) = &intrinsic_fn {
        intrinsic.extend(quote_spanned! {int.span()=>
            self::intrinsic(__wgt__);
        })
    }

    for prop in properties.iter().flat_map(|i| i.properties.iter()) {
        if prop.is_unset() {
            let id = prop.property_id();
            intrinsic.extend(quote! {
                __wgt__.insert_unset(#crate_core::property::Importance::WIDGET, #id);
            });
        } else if prop.value.is_some() {
            let args = prop.args_new();
            intrinsic.extend(quote! {
                __wgt__.insert_property(#crate_core::property::Importance::WIDGET, #args);
            });
        }
    }

    let build = if let Some(build) = &build_fn {
        let out = &build.sig.output;
        let ident = &build.sig.ident;
        quote_spanned! {build.span()=>
            #[doc(hidden)]
            pub fn __build__(__wgt__: #crate_core::property::WidgetBuilder) #out {
                self::#ident(__wgt__)
            }
        }
    } else if let Some(inh) = inherits.last() {
        let path = &inh.path;
        quote! {
            #[doc(hidden)]
            pub use #path::__build__;
        }
    } else {
        errors.push(
            "missing `build(WidgetBuilder) -> T` function, must be provided or inherited",
            ident.span(),
        );
        quote! {
            #[doc(hidden)]
            pub fn __build__(_: #crate_core::property::WidgetBuilder) -> #crate_core::NilUiNode {
                #crate_core::NilUiNode
            }
        }
    };

    let mut inherits_reexport = quote!();
    if let Some(Inherit { attrs, path }) = inherits.first() {
        let mut instructions = quote!();
        for p in &properties {
            for p in &p.removes {
                instructions.extend(quote! {
                    remove { #p }
                });
            }
        }
        for Inherit { attrs, path } in &inherits[1..] {
            instructions.extend(quote! {
                inherit { #path }
            });
        }

        inherits_reexport = quote! {
            #path! {
                >> reexport {
                    #instructions
                }
            }
        }
    }

    let mut widget_macro_intrinsic = quote!();
    for p in properties.iter().flat_map(|p| p.properties.iter()) {
        let cfg = Attributes::new(p.attrs.clone()).cfg;
        let ident = p.property_ident();
        let required = &p.is_required();
        let default = p.has_default();
        widget_macro_intrinsic.extend(quote! {
            property {
                cfg { #cfg }
                ident { #ident }
                required { #required }
                default { #default }
            }
        });
    }

    let macro_ident = ident!("{ident}_{uuid}");

    let r = quote! {
        #(#attrs)*
        #vis mod #ident {
            // custom items
            #(#others)*

            // use items (after custom items in case of custom macro_rules re-export)
            #(#uses)*

            #inherits_reexport

            #intrinsic_fn

            #[doc(hidden)]
            pub fn __intrinsic__(__wgt__: &mut #crate_core::property::WidgetBuilder) {
                #intrinsic
            }

            #build_fn
            #build

            #[doc(hidden)]
            pub mod __core__ {
                pub use #crate_core::{widget_new, widget_reexport};
            }
        }

        #wgt_cfg
        #[doc(hidden)]
        #[macro_export]
        macro_rules! #macro_ident {
            (>> reexport {
                $($instructions:tt)*
            }) => {
                #mod_path::__core__::widget_reexport! {
                    intrinsic {
                        widget { #mod_path }
                        #widget_macro_intrinsic
                    }
                    $($instructions)*
                }
            };
            ($($tt:tt)*) => {
                #mod_path::__core__::widget_new! {
                    intrinsic {
                        widget { #mod_path }
                        #widget_macro_intrinsic
                    }
                    instance {
                        $($tt)*
                    }
                }
            };
        }
        #[doc(hidden)]
        #vis use #macro_ident as #ident;

        #errors
    };
    r.into()
}

struct ArgPath {
    path: TokenStream,
}
impl Parse for ArgPath {
    fn parse(input: parse::ParseStream) -> syn::Result<Self> {
        let fork = input.fork();
        match (fork.parse::<Token![$]>(), fork.parse::<syn::Path>()) {
            (Ok(_), Ok(p)) => {
                if fork.is_empty() {
                    if p.segments[0].ident == "crate" {
                        Ok(ArgPath {
                            path: input.parse().unwrap(),
                        })
                    } else {
                        Err(syn::Error::new(p.segments[0].ident.span(), "expected `crate`"))
                    }
                } else {
                    Err(syn::Error::new(fork.span(), "unexpected token"))
                }
            }
            (Ok(_), Err(e)) => {
                if !util::span_is_call_site(e.span()) {
                    Err(e)
                } else {
                    Err(syn::Error::new(util::last_span(input.parse().unwrap()), e.to_string()))
                }
            }
            _ => Err(syn::Error::new(
                input.span(),
                "expected a macro_rules `$crate` path to this widget mod",
            )),
        }
    }
}

struct WidgetItems {
    uses: Vec<ItemUse>,
    inherits: Vec<Inherit>,
    properties: Vec<Properties>,
    intrinsic_fn: Option<ItemFn>,
    build_fn: Option<ItemFn>,
    others: Vec<Item>,
}
impl WidgetItems {
    fn new(items: Vec<Item>, errors: &mut Errors) -> Self {
        let mut uses = vec![];
        let mut inherits = vec![];
        let mut properties = vec![];
        let mut intrinsic_fn = None;
        let mut build_fn = None;
        let mut others = vec![];

        for item in items {
            match item {
                Item::Use(use_) => {
                    uses.push(use_);
                }
                // match properties!
                Item::Macro(ItemMacro { mac, ident: None, .. }) if mac.path.get_ident().map(|i| i == "properties").unwrap_or(false) => {
                    match syn::parse2::<Properties>(mac.tokens) {
                        Ok(mut p) => {
                            errors.extend(mem::take(&mut p.errors));
                            properties.push(p)
                        }
                        Err(e) => errors.push_syn(e),
                    }
                }
                // match inherit!
                Item::Macro(ItemMacro {
                    mac, attrs, ident: None, ..
                }) if mac.path.get_ident().map(|i| i == "inherit").unwrap_or(false) => match parse2::<Inherit>(mac.tokens) {
                    Ok(mut ps) => {
                        ps.attrs.extend(attrs);
                        inherits.push(ps)
                    }
                    Err(e) => errors.push_syn(e),
                },

                // match fn intrinsic(..)
                Item::Fn(fn_) if fn_.sig.ident == "intrinsic" => {
                    intrinsic_fn = Some(fn_);
                }
                // match fn build(..)
                Item::Fn(fn_) if fn_.sig.ident == "build" => {
                    build_fn = Some(fn_);
                }
                // other user items.
                item => others.push(item),
            }
        }

        WidgetItems {
            uses,
            inherits,
            properties,
            intrinsic_fn,
            build_fn,
            others,
        }
    }
}

struct Inherit {
    attrs: Vec<Attribute>,
    path: Path,
}
impl Parse for Inherit {
    fn parse(input: parse::ParseStream) -> syn::Result<Self> {
        Ok(Inherit {
            attrs: vec![],
            path: input.parse()?,
        })
    }
}

struct Properties {
    errors: Errors,
    properties: Vec<WgtProperty>,
    removes: Vec<Ident>,
    whens: Vec<WgtWhen>,
}

impl Parse for Properties {
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let mut errors = Errors::default();
        let mut properties = vec![];
        let mut removes = vec![];
        let mut whens = vec![];

        while input.is_empty() {
            let attrs = parse_outer_attrs(input, &mut errors);

            if input.peek(widget_util::keyword::when) {
                if let Some(mut when) = WgtWhen::parse(input, &mut errors) {
                    when.attrs = attrs;
                    whens.push(when);
                }
            } else if input.peek(widget_util::keyword::remove) && input.peek2(syn::token::Brace) {
                parse_remove(input, &mut removes, &mut errors);
            } else if input.peek(Ident) || input.peek(Token![crate]) || input.peek(Token![super]) || input.peek(Token![self]) {
                // peek ident or path (including keywords because of super:: and self::). {
                match input.parse::<WgtProperty>() {
                    Ok(mut p) => {
                        p.attrs = attrs;
                        if !input.is_empty() && p.semi.is_none() {
                            errors.push("expected `;`", input.span());
                            while !(input.is_empty()
                                || input.peek(Ident)
                                || input.peek(Token![crate])
                                || input.peek(Token![super])
                                || input.peek(Token![self])
                                || input.peek(Token![#]) && input.peek(token::Bracket))
                            {
                                // skip to next value item.
                                let _ = input.parse::<TokenTree>();
                            }
                        }
                        properties.push(p);
                    }
                    Err(e) => {
                        let (recoverable, e) = e.recoverable();
                        if recoverable {
                            errors.push_syn(e);
                        } else {
                            return Err(e);
                        }
                    }
                }
            } else {
                errors.push("expected `when`, `child`, `remove` or a property declaration", input.span());

                // suppress the "unexpected token" error from syn parse.
                let _ = input.parse::<TokenStream>();

                break;
            }
        }

        Ok(Properties {
            errors,
            properties,
            removes,
            whens,
        })
    }
}

/*
    REEXPORT
*/

pub fn expand_reexport(args: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let args = parse_macro_input!(args as ReexportArgs);

    let r = quote! {};
    r.into()
}

struct ReexportArgs {
    widget: TokenStream,
    intrinsic: Vec<IntrinsicItem>,
    instructions: Vec<ReexportInstruction>,
}
impl Parse for ReexportArgs {
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let intr = non_user_braced!(input, "intrinsic");
        let widget = non_user_braced!(&intr, "widget").parse().unwrap();
        let mut intrinsic = vec![];
        while !intr.is_empty() {
            intrinsic.push(intr.parse().unwrap());
        }

        let inst = non_user_braced!(input, "instructions");
        let mut instructions = vec![];
        while !inst.is_empty() {
            instructions.push(inst.parse().unwrap());
        }

        Ok(ReexportArgs {
            widget,
            intrinsic,
            instructions,
        })
    }
}

enum IntrinsicItem {
    Property {
        cfg: TokenStream,
        ident: Ident,
        required: bool,
        default: bool,
    },
}
impl Parse for IntrinsicItem {
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let input = non_user_braced!(input, "property");
        let r = IntrinsicItem::Property {
            cfg: non_user_braced!(&input, "cfg").parse().unwrap(),
            ident: non_user_braced!(&input, "ident").parse().unwrap_or_else(|e| non_user_error!(e)),
            required: non_user_braced!(&input, "required").parse::<LitBool>()
                .unwrap_or_else(|e| non_user_error!(e))
                .value,
            default: non_user_braced!(&input, "default").parse::<LitBool>()
                .unwrap_or_else(|e| non_user_error!(e))
                .value,
        };
        Ok(r)
    }
}

enum ReexportInstruction {
    Remove(Ident),
    Inherit(TokenStream),
}
impl Parse for ReexportInstruction {
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let tag: Ident = input.parse().unwrap_or_else(|e| non_user_error!(e));
        if tag == "remove" {
            Ok(ReexportInstruction::Remove(
                non_user_braced!(input).parse().unwrap_or_else(|e| non_user_error!(e)),
            ))
        } else if tag == "inherit" {
            Ok(ReexportInstruction::Inherit(non_user_braced!(input).parse().unwrap()))
        } else {
            non_user_error!("{}", tag)
        }
    }
}
