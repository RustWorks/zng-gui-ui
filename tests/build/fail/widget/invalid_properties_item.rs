use zero_ui::core::widget2;

#[widget2($crate::widget)]
pub mod widget {
    properties! {
        {}
    }
}

fn main() {}
