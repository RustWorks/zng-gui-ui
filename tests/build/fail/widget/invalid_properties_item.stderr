error: expected `when`, `child` or a property declaration
 --> $DIR/invalid_properties_item.rs:6:9
  |
6 |         {}
  |         ^^
